programs/locker/src/instructions/v3/claim.rs

```diff
@@ -0,0 +1,224 @@
+use crate::util::{transfer_to_user_v3, MemoTransferContext};
+use crate::*;
+use anchor_spl::memo::Memo;
+use anchor_spl::token_interface::{Mint, TokenAccount, TokenInterface};
+use safe_math::SafeMath;
+use solana_program::hash::hashv;
+use util::{
+    close, parse_remaining_accounts, AccountsType, ParsedRemainingAccounts,
+    TRANSFER_MEMO_CLAIM_VESTING,
+};
+
+const LEAF_PREFIX: &[u8] = &[0];
+const INTERMEDIATE_PREFIX: &[u8] = &[1];
+
+#[derive(AnchorSerialize, AnchorDeserialize)]
+/// Accounts for [locker::create_vesting_escrow].
+pub struct ClaimV3Params {
+    pub vesting_start_time: u64,
+    pub cliff_time: u64,
+    pub frequency: u64,
+    pub cliff_unlock_amount: u64,
+    pub amount_per_period: u64,
+    pub number_of_period: u64,
+    pub max_amount: u64,
+    pub proof: Vec<[u8; 32]>,
+}
+
+impl ClaimV3Params {
+    pub fn get_total_locked_amount(&self) -> Result<u64> {
+        let total_locked_amount = self
+            .cliff_unlock_amount
+            .safe_add(self.amount_per_period.safe_mul(self.number_of_period)?)?;
+        Ok(total_locked_amount)
+    }
+
+    pub fn verify_recipient(&self, recipient: Pubkey, root: [u8; 32]) -> Result<()> {
+        let total_locked_amount = self.get_total_locked_amount()?;
+        let node = hashv(&[
+            &recipient.key().to_bytes(),
+            &total_locked_amount.to_le_bytes(),
+            &self.vesting_start_time.to_le_bytes(),
+            &self.cliff_time.to_le_bytes(),
+            &self.frequency.to_le_bytes(),
+        ]);
+
+        let leaf = hashv(&[LEAF_PREFIX, &node.to_bytes()]);
+
+        let mut computed_hash = leaf.to_bytes();
+        for p in self.proof.iter() {
+            if computed_hash <= *p {
+                computed_hash = hashv(&[&INTERMEDIATE_PREFIX, &computed_hash, p]).to_bytes();
+            } else {
+                computed_hash = hashv(&[&INTERMEDIATE_PREFIX, p, &computed_hash]).to_bytes();
+            }
+        }
+
+        require!(computed_hash == root, LockerError::InvalidMerkleProof);
+
+        Ok(())
+    }
+
+    pub fn get_max_unlocked_amount(&self, current_ts: u64) -> Result<u64> {
+        if current_ts < self.cliff_time {
+            return Ok(0);
+        }
+        let period = current_ts
+            .safe_sub(self.cliff_time)?
+            .safe_div(self.frequency)?;
+        let period = period.min(self.number_of_period);
+
+        let unlocked_amount = self
+            .cliff_unlock_amount
+            .safe_add(period.safe_mul(self.amount_per_period)?)?;
+
+        Ok(unlocked_amount)
+    }
+
+    pub fn get_claimable_amount(&self, total_claimed_amount: u64, current_ts: u64) -> Result<u64> {
```

Comment:
could we make it common function? caused vestingEscrow state also has this function

Reply:
yes, we will make a common function with its logic. Then `get_claimable_amount` in both vestingEscrow and vestingEscrowV3 can use it.

Reply:
But actually, I think we don’t necessarily need to create a common function. Even though the logic is similar, it belongs to two different structs. Let’s keep them independent for simplicity.